File: client/src/test/java/com/jraska/github/client/RealDeepLinkLauncherTest.java
package com.jraska.github.client;

import android.content.Intent;

import com.jraska.github.client.ui.BaseActivity;
import com.jraska.github.client.ui.RepoDetailActivity;
import com.jraska.github.client.ui.UserDetailActivity;
import com.jraska.github.client.ui.UsersActivity;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.ArgumentCaptor;
import org.mockito.Captor;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.robolectric.RobolectricTestRunner;

import java.util.Set;

import okhttp3.HttpUrl;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.verify;

@RunWith(RobolectricTestRunner.class)
public class RealDeepLinkLauncherTest {
  @Captor ArgumentCaptor<Intent> intentCaptor;
  @Mock BaseActivity activity;

  RealDeepLinkLauncher deepLinkLauncher;

  @Before
  public void before() {
    MockitoAnnotations.initMocks(this);

    deepLinkLauncher = new RealDeepLinkLauncher(() -> activity);
  }

  @Test(expected = IllegalArgumentException.class)
  public void whenUnknownHostUrl_thenException() {
    deepLinkLauncher.launch(HttpUrl.parse("https://jraska.com/users"));
  }

  @Test(expected = IllegalArgumentException.class)
  public void whenUnsupportedUrl_thenIllegalStateException() {
    deepLinkLauncher.launch(HttpUrl.parse("https://github.com/jraska/repos/johny"));
  }

  @Test
  public void whenUsersUrl_thenUsersActivityStarts() {
    deepLinkLauncher.launch(HttpUrl.parse("https://github.com/users"));

    verify(activity).startActivity(intentCaptor.capture());
    assertThat(intentCaptor.getValue().getComponent().getClassName())
      .isEqualTo(UsersActivity.class.getName());
  }

  @Test
  public void whenRepoUrl_thenUsersActivityStarts() {
    deepLinkLauncher.launch(HttpUrl.parse("https://github.com/jraska/Falcon"));

    verify(activity).startActivity(intentCaptor.capture());
    assertThat(intentCaptor.getValue().getComponent().getClassName())
      .isEqualTo(RepoDetailActivity.class.getName());
  }

  @Test
  public void whenUserDetailUrl_thenUsersActivityStarts() {
    deepLinkLauncher.launch(HttpUrl.parse("https://github.com/jraska"));

    verify(activity).startActivity(intentCaptor.capture());
    Intent intent = intentCaptor.getValue();
    assertThat(intent.getComponent().getClassName())
      .isEqualTo(UserDetailActivity.class.getName());
    Set<String> keySet = intent.getExtras().keySet();
    for (String key : keySet) {
      assertThat(intent.getStringExtra(key)).isEqualTo("jraska");
    }
  }
}

File: client/src/test/java/com/jraska/github/client/RealDeepLinkLauncherTest.kt
package com.jraska.github.client

import android.content.Intent
import com.jraska.github.client.ui.BaseActivity
import okhttp3.HttpUrl
import org.junit.Before
import org.junit.Test
import org.mockito.ArgumentMatchers.any
import org.mockito.Mock
import org.mockito.Mockito.verify
import org.mockito.MockitoAnnotations
import javax.inject.Provider

class RealDeepLinkLauncherTest {
  @Mock private lateinit var activity: BaseActivity
  private lateinit var deepLinkLauncher: RealDeepLinkLauncher

  @Before
  fun before() {
    MockitoAnnotations.initMocks(this)

    deepLinkLauncher = RealDeepLinkLauncher(Provider { activity })
  }

  @Test(expected = IllegalArgumentException::class)
  fun whenUnknownHostUrl_thenException() {
    deepLinkLauncher.launch(HttpUrl.get("https://jraska.com/users"))
  }

  @Test(expected = IllegalArgumentException::class)
  fun whenUnsupportedUrl_thenIllegalStateException() {
    deepLinkLauncher.launch(HttpUrl.get("https://github.com/jraska/repos/johny"))
  }

  @Test
  fun whenUsersUrl_thenPasses() {
    deepLinkLauncher.launch(HttpUrl.get("https://github.com/users"))

    verify<BaseActivity>(activity).startActivity(any<Intent>())
  }

  @Test
  fun whenRepoUrl_thenPasses() {
    deepLinkLauncher.launch(HttpUrl.get("https://github.com/jraska/Falcon"))

    verify<BaseActivity>(activity).startActivity(any<Intent>())
  }

  @Test
  fun whenUserDetailUrl_thenUsersActivityStarts() {
    deepLinkLauncher.launch(HttpUrl.get("https://github.com/jraska"))

    verify<BaseActivity>(activity).startActivity(any<Intent>())
  }
}
